test-interface/src/main/scala/org/scalajs/testinterface/internal/Slave.scala

```diff
@@ -5,119 +5,70 @@ import js.annotation._
-
-  private def sendOutboundRunnerMessage(msg: String): Unit = {
-    assert(canSendRunnerMessage)
-    Com.send(s"msg:$msg")
-  }
-
-  private def allowSendRunnerMessage[T](body: => T): T = {
-    try {
-      canSendRunnerMessage = true
-
-      // Flush the queue
-      while (!messageQueue.isEmpty)
-        sendOutboundRunnerMessage(messageQueue.dequeue)
-
-      body
-    } finally {
-      canSendRunnerMessage = false
-    }
-  }
+  JSRPC.attach(JSSlaveEndpoints.newRunner)(newRunner _)
+  JSRPC.attachAsync(JSSlaveEndpoints.execute)(execute _)
+  JSRPC.attach(JSSlaveEndpoints.stopSlave)(stopSlave _)
+  JSRPC.attach(JSSlaveEndpoints.msg)(runner.receiveMessage _)
 
   // Message handler methods
 
-  private def newRunner(): Try[Unit] = {
+  private def newRunner(req: Unit): Unit = {
+    val framework = FrameworkLoader.loadFramework(frameworkName)
     val loader = new ScalaJSClassLoader(
         scala.scalajs.runtime.environmentInfo.exportsNamespace)
-    Try(runner = framework.slaveRunner(args.toArray, remoteArgs.toArray,
-        loader, outboundRunnerMessage))
+    runner = framework.slaveRunner(args.toArray, remoteArgs.toArray,
+        loader, JSRPC.send(JVMSlaveEndpoints.msg))
   }
 
-  private def execute(data: js.Dynamic): Unit = {
+  private def execute(req: ExecuteRequest): Future[List[TaskInfo]] = {
     ensureRunnerExists()
 
-    val sTask = data.serializedTask.asInstanceOf[String]
-    val task = runner.deserializeTask(sTask, str =>
-      TaskDefSerializer.deserialize(js.JSON.parse(str)))
-
+    val task = TaskInfoBuilder.attachTask(req.taskInfo, runner)
     val eventHandler = new RemoteEventHandler
 
-    val colorSupport = data.loggerColorSupport.asInstanceOf[js.Array[Boolean]]
     val loggers = for {
-      (withColor, i) <- colorSupport.zipWithIndex
+      (withColor, i) <- req.loggerColorSupport.zipWithIndex
     } yield new RemoteLogger(i, withColor)
 
+    val promise = Promise[List[TaskInfo]]
+
     def cont(tasks: Array[Task]) = {
-      val result = Try(js.JSON.stringify(tasks2TaskInfos(tasks, runner)))
+      val result = Try(tasks.map(TaskInfoBuilder.detachTask(_, runner)).toList)
       eventHandler.invalidate()
       loggers.foreach(_.invalidate())
-      reply(result)
+      promise.complete(result)
     }
 
-    val launched = Try(task.execute(eventHandler, loggers.toArray, cont))
-
-    if (launched.isFailure)
-      reply(launched)
-  }
+    try {
+      task.execute(eventHandler, loggers.toArray, cont)
+    } catch {
+      case NonFatal(t) =>
+        promise.tryFailure(t)
+    }
 
-  private def stopSlave(): Try[Unit] = {
-    ensureRunnerExists()
-    val res = Try { runner.done(); () }
-    runner = null
-    res
+    promise.future
   }
 
-  private def incomingRunnerMessage(msg: String): Try[Unit] = {
+  private def stopSlave(req: Unit): Unit = {
     ensureRunnerExists()
-    Try { runner.receiveMessage(msg); () }
+    try runner.done()
+    finally runner = null
```

Comment:
Seems like either of two things hold here:
* either it is not valid to send a message to a slave after `done()` has been called on it, in which case sbt-plugin-test should not try to do that, or
* we should simply not set `runner` to `null` here, and instead just let it live on.